src/main/java/org/mockito/internal/configuration/plugins/PluginRegistry.java

```diff
@@ -46,8 +48,8 @@ class PluginRegistry {
     private final List<MockResolver> mockResolvers =
             new PluginLoader(pluginSwitch).loadPlugins(MockResolver.class);
 
-    private final DoNotMockEnforcer doNotMockEnforcer =
-            new PluginLoader(pluginSwitch).loadPlugin(DoNotMockEnforcer.class);
+    private final DoNotMockEnforcerWithType doNotMockEnforcer = (DoNotMockEnforcerWithType)
+            (new PluginLoader(pluginSwitch).loadPlugin(DoNotMockEnforcerWithType.class, DoNotMockEnforcer.class));
```

Comment:
We should probably update `loadPlugin` to become the following to avoid the cast:

```
<PreferredT, AlternateType extends PreferredT> PreferredT loadPlugin(
            final Class<PreferredT> preferredPluginType,
            final Class<AlternateType> alternatePluginType)
```
This then also allows us to remove the `@SuppressWarnings("unchecked")`. Let's do that in a follow-up?

Reply:
Yup. Sure thing.

Reply:
@TimvdLippe actually, I can do that to remove the ugliness from the caller, but the other method will still require an unchecked type cast either way:

```diff
diff --git a/src/main/java/org/mockito/internal/configuration/plugins/PluginLoader.java b/src/main/java/org/mockito/internal/configuration/plugins/PluginLoader.java
index e55417332..19ea08ed9 100644
--- a/src/main/java/org/mockito/internal/configuration/plugins/PluginLoader.java
+++ b/src/main/java/org/mockito/internal/configuration/plugins/PluginLoader.java
@@ -58,7 +58,7 @@ class PluginLoader {
      * @return An object of either {@code preferredPluginType} or {@code alternatePluginType}
      */
     @SuppressWarnings("unchecked")
-    <PreferredT, AlternateType> Object loadPlugin(
+    <PreferredT, AlternateType extends PreferredT> PreferredT loadPlugin(
             final Class<PreferredT> preferredPluginType,
             final Class<AlternateType> alternatePluginType) {
         try {
@@ -74,7 +74,7 @@ class PluginLoader {
 
             return plugins.getDefaultPlugin(preferredPluginType);
         } catch (final Throwable t) {
-            return Proxy.newProxyInstance(
+            return (PreferredT) Proxy.newProxyInstance(
                     preferredPluginType.getClassLoader(),
                     new Class<?>[] {preferredPluginType},
                     new InvocationHandler() {
```